cocos/animation/core/shared-stack-based-allocator.ts

```diff
@@ -0,0 +1,299 @@
+import { DEBUG, TEST } from 'internal:constants';
+import { assertIsTrue, binarySearchEpsilon } from '../../core';
+
+const allocatorPageCountTag = Symbol(DEBUG ? '[[The count of pages used by this allocator.]]' : '');
+
+const onStackPurgedTag = Symbol(DEBUG ? `[[Notify that theres is no allocator on the stack anymore.]]` : '');
+
+class SharedMemoryPage {
+    constructor (byteLength: number) {
+        this.buffer = new ArrayBuffer(byteLength);
+    }
+
+    public readonly buffer: ArrayBuffer;
+
+    public useCount = 0;
+}
+
+class PagedStack {
+    constructor (private _up: SharedStackBasedAllocatorManager, private _pageSize: number) {
+    }
+
+    get pageSize () {
+        return this._pageSize;
+    }
+
+    get debugLocking () {
+        return this._locking;
+    }
+
+    public get allocatorCount () {
+        return this._allocatorCount;
+    }
+
+    public debugLock () {
+        assertIsTrue(!this._locking, `The memory is locking.`);
+        this._locking = true;
+    }
+
+    public debugUnlock () {
+        assertIsTrue(this._locking, `Wrong execution logic: the memory is not locking.`);
+        this._locking = false;
+    }
+
+    public getPageMemory (index: number) {
+        assertIsTrue(index >= 0 && index < this._pages.length, `Page index out of range`);
+        return this._pages[index].buffer;
+    }
+
+    public pushPage (allocator: SharedStackBasedAllocator): SharedMemoryPage {
+        const oldAllocatorPageCount = allocator[allocatorPageCountTag];
```

Comment:
The `PagedStack` is shared by multiple allocators. A `pushPage()` request:

- may really push a page onto the stack - if current page count is not sufficient to satisfy the requesting allocator.

- or may do not involve page creation since current  page count is enough.

For example, in a time, the paged stack and allocators may be like:

```
Stack      |  ·····>  (real stack,    5 pages)
Allocator1 |  ·····>  (virtual stack, 5 pages)
Allocator2 |  ···>    (virtual stack, 3 pages)
```

In `Stack`, page 0,1,2 has two allocator users, while page 3,4 has only one allocator user.

Now a `pushPage()` is requested from `Allocator2`. Since `Allocator2` needs 4 pages and the stack is sufficient, the stack does not grow. Only some counter is updated:

```
Stack      |  ·····>  (real stack,    5 pages)
Allocator1 |  ·····>  (virtual stack, 5 pages)
Allocator2 |  ····>   (virtual stack, 4 pages)
```

Then, a `pushPage()` is requested from `Allocator1`. Since `Allocator1` needs 5 pages and the stack is insufficient, the stack needs to grow:

```
Stack      |  ······> (real stack,    6 pages)
Allocator1 |  ······> (virtual stack, 6 pages)
Allocator2 |  ····>   (virtual stack, 4 pages)
```

As you can see from above: **the stack have to know which allocator is firing the incoming `pushPage()` request**, and calculates its required pages(simply get its count and increase by 1).